### ISSUE TITLE
Not iterating over all files

### ISSUE THREAD

**Issue by pgoostree** [mariusGundersen/gulp-flatMap#11]
I have a task that looks like this below.  There are two files in the src folder in the gulp task, however the generateStyleSheet() method is only being called once.  Do you have any tips on why this is?  

Thanks, 

Peter

```
gulp.task('compile-less', function () {
    gulp.src("./Styles/less/theme/properties/*.less")
        .pipe(foreach(function (stream, file) {
            generateStyleSheet(stream, file);
        }))
});


function generateStyleSheet(stream, file) {
    gulp.src('./Styles/less/theme/template.less')
        .pipe(inject(stream, {
            starttag: '/* inject:imports */',
            endtag: '/* endinject */',
            transform: function (filepath) {
                return '@import ".' + filepath + '";';
            }
        }))
        .pipe(less())
        .pipe(rename({
            suffix: ".g",
            basename: path.basename(file.path, '.less')
        }))
        .pipe(gulp.dest('./Styles/css/themes'));
}
```

**Comment by pgoostree** [comment]
I figured out what my issue is.  I am not returning the stream to the foreach handler.  That seemed to have fixed my problem.

**Comment by mariusGundersen** [comment]
I've had the same problem, with the same fix as you. It is a problem that it doesn't fail with an error or warning, so it's definitely a bug.

**Comment by yattias** [comment]
+1

**Comment by Gerst20051** [comment]
:+1: i'm returning stream but still not iterating over every file

**Comment by nbcormier** [comment]
I am also returning the stream, but it is not iterating over every file.  It seems to stop after exactly 16 files.

**Comment by iggyvanlith** [comment]
+1 Some problem here with not iterating over every file. Seems to stop after 16 files like @nbcormier said.

**Comment by mariusGundersen** [comment]
Can one of @yattias, @Gerst20051, @nbcormier or @iggyvanlith post a small test code that demonstrates the problem?

**Comment by nbcormier** [comment]
Sure.  To be clear, in total, it may end up iterating over more than 16 files.  However, it will only iterate through a depth of 16 for each directory. So, if my theory is correct, your base directory could have 16 children, with each of those having 16 children, and so on...and you would probably iterate over everything.  Hope that makes sense.

```js
        function uploadDir(dir) {
            log('Uploading ' + dir);

            gulp.src(dir + '**/*.*')
            .pipe(foreach(function(stream, file) {
                uploadFile(file);
                return stream;
            }));

            function uploadFile(file) {
               // do stuff to upload file
            }
        }
```

**Comment by mariusGundersen** [comment]
@nbcormier, you should probably use the [gulp-tap](https://www.npmjs.com/package/gulp-tap) plugin rather than this one. This plugin is unfortunately misnamed, it behave more like flatMap than foreach, so it has a very specific use case.

**Comment by nbcormier** [comment]
Thanks @mariusGundersen, gulp-tap does iterate through all the files in my glob.

**Comment by RobeeeJay** [comment]
I hit this issue too, but using the https://www.npmjs.com/package/glob module was a better fit for me than gulp-tap. Felt better using that than creating streams.

**Comment by mariusGundersen** [comment]
I've renamed this project into gulp-flatmap, since that is actually what it does. I've also updated it to require returning a stream, which should prevent the issue reported in this issue (there is now a unit test that confirms that it works for 32 files. 

So, either install gulp-tap, if you want to do something for each file in the stream, or install gulp-flatmap if you want to convert each file in the stream into multiple files.
